Changes based on feedback by Electric Imp

sumeetrajpal opened the pull request:
-Updated version format (Also changed version number from 1.0.1 to 1.1.0 as changes to agent constructer are not backward compatible)
-Incorporated changes made in Electric imps master branch to the Readme.md file
-Included hyperlink to sign up page in setup instructions in Readme.md
-Tested email verification link not clickable bug has been fixed on Conctr
-Included more in depth instructions on where to find required data for constructor.
-Changed agent constructor to accept a Table of options as an optional parameter 
-Fixed inconsistency with setOpts and constructor arguments.

sumeetrajpal committed:
*sendData() now accepts a array of payloads.
*Posts an array to Conctr vs singular payloads.
*Sanity check on timestamps in the agent.

sumeetrajpal committed:
Updated version number

sumeetrajpal committed:
Fixes from previous pull request.

Commit:
Fixed timestamp and location issues, timeout checks and cleaned up logic.

sumeetrajpal committed:
Merge pull request #1 from blindman2k/master

Fixed timestamp and location, timeout checks and cleaned up logic

sumeetrajpal committed:
Updated timestamp sanity check to check for a string of numbers.

sumeetrajpal committed:
turned off debug mode

sumeetrajpal committed:
Made changes requested by imp

sumeetrajpal committed:
minor fix

blindman2k commented on README.md:
```diff
@@ -2,33 +2,34 @@
-The constructor takes three required parameters: your application id, API key and model. There is also  three optional parameter, the region to be used (defaults to us-west-2), the environment (defaults to staging) and the useAgentId (defaults to false).
+### Constructor: Conctr(*appId, apiKey, model[, options]*)
+
+The constructor takes three required parameters: your application ID, API key and model. These details can be found by navigating into your application on the Conctr platform, selecting on the *models* tab in the left side menu then click on the example button under the model you wish to use and chose the tab marked *Squirell*. There are also three optional parameters: the region to be used (defa ...
```
"*models* tab" should be "*Models* tab" and "example button" should be "*Example* button".
"*Squirell*" should be "*Squirrel*". "options" should be "*Options*".

blindman2k commented on README.md:
````diff
@@ -2,33 +2,34 @@
 
-##### Example
 ```squirrel
 #require "conctr.agent.class.nut:1.0.0"
````
Need to update the version number.

blindman2k commented on README.md:
```diff
@@ -41,112 +42,98 @@ conctr <- Conctr(APP_ID, API_KEY, MODEL);
+conctr.sendData(currentTempAndPressure, function(error, response) {
     if (error) {
-        //handle error
+        // Handle error
```
Remove this comment and replace with an error log.
server.error("Failed to deliver to Conctr: " + error);

blindman2k commented on README.md:
```diff
@@ -41,112 +42,98 @@ conctr <- Conctr(APP_ID, API_KEY, MODEL);
+        // Handle error
     } else {
-        //data was successfully recieved by Conctr
+        server.log("Data was successfully recieved by Conctr");
```
Replace "recieved" with "received".

blindman2k commented on README.md:
````diff
@@ -41,112 +42,98 @@ conctr <- Conctr(APP_ID, API_KEY, MODEL);
+#### Example
+
 ```squirrel
 #require "conctr.device.class.nut:1.0.0"
````
Update version number.

blindman2k commented on README.md:
```diff
@@ -41,112 +42,98 @@ conctr <- Conctr(APP_ID, API_KEY, MODEL);
+conctr.sendData(currentTempAndPressure, function(error, response) {
     if (error) {
-        //handle error
+        // Handle error
```
Similar error log here.

blindman2k commented on README.md:
```diff
@@ -41,112 +42,98 @@ conctr <- Conctr(APP_ID, API_KEY, MODEL);
+        // Handle error
     } else {
-        //data was successfully recieved by Conctr
+        server.log("Data was successfully recieved from the device by Conctr");
```
Replace "recieved" with "received".

blindman2k commented on conctr.agent.class.nut:
```diff
@@ -92,13 +95,13 @@ class Conctr {
         // If it's a table, make it an array
         if (typeof payload == "table") {
-            payload = [ payload ];
+            payload = [payload];
```
The spaces where better. Not sure why you took them out again.

sumeetrajpal replied:
my auto formatter :) will update it to stop doing that.

blindman2k commented on conctr.agent.class.nut:
```diff
@@ -7,13 +6,14 @@
     static SOURCE_AGENT = "impagent";
+    static MIN_TIME = 946684801; // Epoch timestamp for 00:01 AM 01/01/2000 (used for timestamp sanity check)
 
     static HOUR_SEC = 3600; // One hour in seconds
```
Rename HOUR_SEC to something like DEFAULT_LOC_INTERVAL.

sumeetrajpal committed:
CONCTR-358 #comment fixes requested by aron

sumeetrajpal commented:
@blindman2k made requested changes

sumeetrajpal committed:
added send function as an alias for the sendData function, it accepts a arguments similar to agent.send()

sumeetrajpal committed:
Minor formatting change

sumeetrajpal commented:
@blindman2k Made requested changes and added a send method to the device class which will work with the same arguments as the agent.send function. Added the description to the readme for this as well.

blindman2k commented on README.md:
````diff
@@ -134,6 +134,32 @@ conctr.sendData(currentTempAndPressure, function(error, response) {
+```squirrel
+local currentTempAndPressure = { "temperature" : 29, "pressure" : 1032};
+
+conctr.send("any string", currentTempAndPressure, function(error, response) {
````
Take the callback out of this example but leave it in otherwise.
agent.send() doesn't support callbacks so it will be confusing.

blindman2k commented on conctr.device.class.nut:
```diff
@@ -145,6 +145,19 @@ class Conctr {
+     * @param  {{Function (err,response)} callback - Callback function on resp from Conctr through agent
+     */
+    function send(unusedKey, payload = null, callback = null) {
+        if ((typeof unusedKey == "table" || typeof unusedKey == "array") && (payload == null || typeof payload == "function")) {
```
We don't need these checks and swaps. Just call sendData.

blindman2k commented on conctr.device.class.nut:
```diff
@@ -6,15 +6,15 @@
 
-    // 1 hour in milliseconds
+    // 1 hour in seconds
     static HOUR_SEC = 3600;
```
Replace HOUR_SEC with DEFAULT_LOC_INTERVAL as you did with the agent.

blindman2k reviewed (changes requested):
Minor tweaks left.

sumeetrajpal committed:
minor tweaks

sumeetrajpal commented:
@blindman2k made minor tweaks

sumeetrajpal committed:
Minor bug fix

sumeetrajpal committed:
Merge branch 'master' into master

sumeetrajpal commented:
@blindman2k resolved the conflicts. Can i merge it in?

sumeetrajpal committed:
Updated version number:
removed callback queue and added Bullwinkle, Message Manager and imp pager support.
added device claiming endpoint
added retries to sending http requests
added new location sending criteria to allow sending of location on a specific wake reason.
updated readme

sumeetrajpal committed:
updated getLocationOpts in device and readme

sumeetrajpal committed:
minor comment update

sumeetrajpal committed:
Merge branch 'release/v2.0.0'

sumeetrajpal committed:
changes based on arons review. More coming.

sumeetrajpal committed:
Changes with aron

sumeetrajpal committed:
added queueing for failed recoverable requests

sumeetrajpal committed:
updated readme and added licence to the top of lib files

sumeetrajpal committed:
made the agent class default to not sending location to make it completely standalone from the device class. updated readme.

sumeetrajpal committed:
added tests and examples

sumeetrajpal committed:
added files missed in last commit, same commit message.

sumeetrajpal committed:
added a manual instantiation of the conctr class as the agent and device files

sumeetrajpal committed:
added more tests

sumeetrajpal committed:
Merge branch 'feature/sumeet.use_instantiated_test_files' into develop

sumeetrajpal committed:
added readmes to examples

sumeetrajpal committed:
Removed credentials. Deleted credentials from conctr as well for security.

sumeetrajpal committed:
Merge branch 'develop'

marcstringer commented on README.md:
```diff
@@ -41,112 +43,120 @@ conctr <- Conctr(APP_ID, API_KEY, MODEL);
+| Key | Data Type | Required | Description |
+| --- | --------- | -------- | ----------- |
+| *payload* | Table/Array | Yes | A table or array of tables containing the data to be sent to Conctr. The keys in the table should correspond to fields from the model and the keys should be of type specified in the model. |
+| *callback* | Function | No | Function to be called on response from Conctr. The function should take two arguements, *error* and *response*. When no error occurred, the first arguement will be null. |
```
should probably have a callback table.
See: https://github.com/mysticpants/Wiznet_5500
e.g
This method triggers the supplied callback function when data is received. The callback takes the following parameters:
Key            | Data Type|Description 
----------  | ---------- | ----------
error           | String .     | An error message
response   |

marcstringer commented on README.md:
```diff
@@ -41,112 +43,120 @@ conctr <- Conctr(APP_ID, API_KEY, MODEL);
+
+| Key | Data Type | Default Value | Description |
+| --- | --------- | ------------- | ----------- |
+| *options.locEnabled* | Boolean | `true` | When enabled, location data will be automatically included with the data payload. |
```
inconsistent with Description ending in a full stop. I think it looks better without a full stop

marcstringer commented on README.md:
```diff
@@ -41,112 +43,120 @@ conctr <- Conctr(APP_ID, API_KEY, MODEL);
+// change options to disable location sending altogether
+local opts = { 
+    "locEnabled" : false,
+    };
```
move the }; left most to align

marcstringer commented on conctr.device.lib.nut:
```diff
@@ -0,0 +1,256 @@
+    // 1 hour in seconds
+    static DEFAULT_LOC_INTERVAL = 3600;
+
+    // Location recording parameters
```
Imp will probably want a comment description per parameter

marcstringer commented on conctr.device.lib.nut:
```diff
@@ -0,0 +1,256 @@
+    }
+
+
+    // 
```
add a description of what the function does

marcstringer commented on conctr.device.lib.nut:
```diff
@@ -0,0 +1,256 @@
+    // 
+    // @param  {Table} options - Table containing options to be sent to the agent
+    // 
+    function _sendAgentOpts(opts) {
```
function description

marcstringer commented on example/send_data_from_agent/agent.example.nut:
```diff
@@ -0,0 +1,47 @@
+conctr <- Conctr(APP_ID, API_KEY, MODEL, conctrOpts);
+
+
+// Example payload. This must contain fields relating to your conctr model.
```
rephrase:
This must contain the fields in your conctr model.

marcstringer commented on example/send_data_from_device/device.example.nut:
```diff
@@ -0,0 +1,56 @@
+}
+
+// Initialize Libs
+bull <- Bullwinkle(bullOpts);
```
Should +bull <- Bullwinkle(bullOpts);
go before 

+conctrOpts <- {
 +    "locEnabled": true,
 +    "locInterval": 3600,
 +    "locSendOnce": false,
 +    "locWakeReasons": [WAKEREASON_BLINKUP, WAKEREASON_PIN, WAKEREASON_NEW_SQUIRREL],
 +    "messageManager": bull
 +}

marcstringer commented on tests/agent.test.nut:
```diff
@@ -0,0 +1,83 @@
+
+    function testSendData() {
+        return Promise(function(resolve, reject) {
+            // Ensure this payload matches your model
```
rephrase // Ensure this payload matches your model
to 
// Ensure the payload fields match your model fields

marcstringer commented on tests/agent.test.nut:
```diff
@@ -0,0 +1,83 @@
+        return "Hi from #{__FILE__}!";
+    }
+
+    function testSendData() {
```
you will want a header message explaining in a line what the test is doing.
e.g
// tests the sendData function receives the correct http response

marcstringer commented on tests/agent.test.nut:
```diff
@@ -0,0 +1,83 @@
+                if (err) reject(err);
+
+                // assert the data was accepted
+                this.assertEqual(201, resp.statuscode);
```
201 is a magic number. should declare it at the top

marcstringer commented on tests/agent.test.nut:
```diff
@@ -0,0 +1,83 @@
+        return Promise(function(resolve, reject) {
+            // Ensure this payload matches your model
+            local payload = {
+                "temperature": 15,
```
worth also making the values of temperature i.e 15 a global constant so can be easily changed.

marcstringer commented on tests/agent.test.nut:
```diff
@@ -0,0 +1,83 @@
+        }.bindenv(this))
+    }
+
+    function testSetDeviceId() {
```
header description

marcstringer commented on tests/agent.test.nut:
```diff
@@ -0,0 +1,83 @@
+            local newDeviceId = "testDevice";
+
+            // Ensure that the device ids do not already match
+            this.assertTrue(conctr._device_id != newDeviceId);
```
is there a better test you could do here?
i.e what should conctr._devicde_id equal?

marcstringer commented on tests/agent.test.nut:
```diff
@@ -0,0 +1,83 @@
+            local req = http.request("POST", endpoint, {}, http.jsonencode({}));
+
+            req.sendasync(function(resp) {
+                this.assertEqual(401, resp.statuscode);
```
magic number 401

marcstringer commented on tests/agent.test.nut:
```diff
@@ -0,0 +1,83 @@
+        }.bindenv(this))
+    }
+
+    function xtestRockyEndpoints() {
```
header. this 1 in particular as there are no comments through it

marcstringer commented on tests/build_files/agent.in.nut:
```diff
@@ -0,0 +1,570 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// #require "#require "messagemanager.class.nut:1.0.2"
```
should this line be here?

marcstringer commented on tests/build_files/device.in.nut:
```diff
@@ -0,0 +1,280 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// "#require "messagemanager.class.nut:1.0.2"
```
delete line? 
 +// "#require "messagemanager.class.nut:1.0.2"

marcstringer commented on tests/device.test.nut:
```diff
@@ -0,0 +1,121 @@
+    }
+
+
+    function testSendData() {
```
Header function description + magic numbers

marcstringer commented on tests/device.test.nut:
```diff
@@ -0,0 +1,121 @@
+                // Check if sender has reply capability
+                if ("onReply" in conctr._sender) {
+                    // assert the data was accepted
+                    this.assertEqual(201, resp.statuscode);
```
assertions in a promise. need to try and catch otherwise will crash the test

marcstringer commented on tests/device.test.nut:
```diff
@@ -0,0 +1,121 @@
+        }.bindenv(this))
+    }
+
+    function testSend() {
```
header + magic numbers + assertions in a promise

marcstringer commented on tests/device.test.nut:
```diff
@@ -0,0 +1,121 @@
+        }.bindenv(this))
+    }
+
+    function testSetLocationSendingOpts() {
```
header function description

marcstringer replied:
+ assertions in a promise

marcstringer commented on tests/device.test.nut:
```diff
@@ -0,0 +1,121 @@
+    }
+
+
+    function xtestMessageManagerReply() {
```
function header message + magic numbers + assertions in a promise

sumeetrajpal committed:
Updated agent opts sending mechanism

sumeetrajpal committed:
pending testing

sumeetrajpal committed:
updated default params and readme

sumeetrajpal committed:
Merge branch 'feature/update_default_location_params' into develop

sumeetrajpal committed:
updates based on marcs feedback

sumeetrajpal committed:
Merge branch 'develop'

marcstringer commented on README.md:
```diff
@@ -2,151 +2,149 @@
 
 The Conctr library allows you to easily integrate your agent and device code with the [Conctr IoT Platform](https://conctr.com). This library provides an easy way to send data to a Conctr application. 
```
Worth writing a couple of lines about what conctr is.

marcstringer commented on README.md:
````diff
@@ -2,151 +2,149 @@
 }.bindenv(this));
 ```
 
+### send(*ignoredString, payload*)
````
Example for send ?

marcstringer commented on conctr.agent.lib.nut:
```diff
@@ -0,0 +1,556 @@
+
+
+    // Conctr Variables
+    _api_key = null;
```
// Conctr Variables,  // Pending queue status, // Location state should have a single line comment

marcstringer commented on conctr.agent.lib.nut:
```diff
@@ -0,0 +1,556 @@
+    // 
+    // @param {Array}       payload    The data payload
+    // @param {Function}    cb         Optional callback
+    // 
```
This header doesnt have a return

marcstringer commented on conctr.agent.lib.nut:
```diff
@@ -0,0 +1,556 @@
+
+
+    // 
+    // Sends a request to the device to send its current location (array of wifis) if conditions in current location sending opts are met. 
```
return in header

marcstringer commented on conctr.agent.lib.nut:
```diff
@@ -0,0 +1,556 @@
+
+
+    // 
+    // Claims a device for a consumer
```
return header

marcstringer commented on conctr.agent.lib.nut:
```diff
@@ -0,0 +1,556 @@
+    // Handles device claim response from Conctr
+    // 
+    // @param  {Object} context Rocky context
+    // 
```
return header

marcstringer commented on conctr.device.lib.nut:
```diff
@@ -0,0 +1,272 @@
+    _locWakeReasons = null; // Array of hardware.wakereasons() 
+
+    // Location state
+    _locSent = false;
```
single likne comment for locsent and locTimeout

marcstringer commented on conctr.agent.lib.nut:
```diff
@@ -0,0 +1,556 @@
+    // Set device unique identifier
+    // 
+    // @param {String} deviceId - Unique identifier for associated device. (Defaults to imp device id)
+    // 
```
return header this

Commit:
addressed comments

Commit:
added a tests readme.md

Commit:
updated test instructions

Commit:
improved readme.md

Commit:
test read me updated

Commit:
added failing test cases

Commit:
specified the location of the agent and device file

Commit:
syntax update

Commit:
fixed test i added

sumeetrajpal committed:
Merge pull request #3 from marcstringer/master

Finished Squirrel Review

sumeetrajpal committed:
updated docs to new require structure

sumeetrajpal committed:
updated device requires as well

sumeetrajpal committed:
updated readme

blindman2k committed:
Simplified location on the agent greatly.

blindman2k committed:
Merge branch 'master' of github.com:sumeetrajpal/squirrel into feature/simpler-agent-location

blindman2k committed:
Changed error reporting in sendData when not using messageManager

sumeetrajpal committed:
updated readme

sumeetrajpal committed:
documented DEBUG mode

blindman2k committed:
Removed setting of default options when no option is provided.

blindman2k committed:
Improved readme for location options

blindman2k committed:
More readme

blindman2k committed:
Merged two claim functions into one
